### ISSUE TITLE
Multi-threaded locking issue and scope errors

### ISSUE THREAD

**Issue by srbaber** [liquibase/liquibase#2966]
## Environment

**Liquibase Version**: 4.11.0

**Liquibase Integration & Version**: startup servlet

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: Oracle 12

**Operating System Type & Version**: centos 7  

**Infrastructure Type/Provider**: local

## Description

A clear and concise description of the issue being addressed.
Exception during liquibase update in a jboss eap7 enviorment with 8 startup threads.
liquibase.exception.LiquibaseException: java.lang.RuntimeException: Cannot end scope kyvpupvnoy when currently at scope ghinssjhuh

## Steps To Reproduce
Since it's a threading issue I don't have a test case to prove the issue

List the steps to reproduce the behavior.
1) Using a debugger, set a breakpoint in the liquibase class inside the runInScope.
    public void update(Contexts contexts, LabelExpression labelExpression, boolean checkLiquibaseTables) throws LiquibaseException {
        this.runInScope(() -> {
            LockService lockService = LockServiceFactory.getInstance().getLockService(this.database);
            lockService.waitForLock();    // breakpoint here for example

2) Start Thread 1 that invokes the liquibase.update and run to the breakpoint
3) Start Thread 2 that invokes the liquibase update and run to the breakpoint
4) Continue Thread 1 to completion

Since Thread 2 has started the runInScope and invoked Scope.child and started a new scope, when the Thread 1 completes it will throw an exception.

## Actual Behavior
The liquibase update for this table fails with the exception and the war fails to deploy.

## Expected/Desired Behavior
The locking of the liquibase update is "more thread safe" so that the exception doesn't happen and the wars are deployed

## Additional Context
I've been able to mitigate the issue by creating a lock service that uses a different lock table, that I invoke around the liquibase.update call.
                lockService.waitForLock();
                liquibase.update(new Contexts(config.getContexts()), new LabelExpression(), checkLiquibaseTables());
                lockService.releaseLock();
So that the locking and single threaded nature occurs outside of the runInScope.

**Comment by loicoudot** [comment]
Hi, this is a regression.
With liquibase 3.8.9 we can do multithread database creation.
With liquibase 4.9.1 we encoutered regulary the same error.

We really appreciate an multithread safe liquibase.

**Comment by asolntsev** [comment]
@srbaber @loicoudot @nvoxland  I managed to stably reproduce the issue. Look at my sample project https://github.com/asolntsev/liquibase-concurrent-run

Checkout the project and run `./gradlew test`. 

The root cause is class `SingletonScopeManager`: it's singleton. You can create many instances of `LiquiBase` class, but they all will still use the same instance of `SingletonScopeManager`. :(

**Comment by asolntsev** [comment]
@srbaber @loicoudot  @nvoxland ^^^ Is anyone alive here? ^^^

**Comment by loicoudot** [comment]
Thank’s for the analysis. I confirm that the error we get in our case is the same. The singleton is not thread safe.

**Comment by asolntsev** [comment]
@loicoudot But how to fix it? I can prepare a pull request, but... I see `Scope` class is used in many other LB classes.

**Comment by bvremmeinfor** [comment]
FYI - another sample project can be found here https://github.com/bvremmeinfor/liquibase-threading-demo

It is reasonable thread safe as it replaces some pluggable parts with thread safe implementations (based on suggestions in other issues). It still fails from time to time in SqlGeneratorFactory. There is a testLoop.sh script for executing multiple tests - must start new JVM for each as remaining issue is SqlGeneratorInit caching.

**Comment by kataggart** [comment]
Hey all we believe this will likely be resolved when https://github.com/liquibase/liquibase/pull/3240 is completed. 

Let us know if you are interested in trying out a snapshot build to make sure it does resolve things. Thanks!

**Comment by asolntsev** [comment]
@kataggart yes, sure, I want to test the snapshot. 
P.S. Why not make thread-safe implementation the default one? It doesn't make anything worse. 

NB! I would say using ThreadLocal is rather a hack. A really God solution would be to avoid using static methods like Scope.getCurrentScope() at all. A new I stance of Scope should be created in constructor `new LiquiBase()`. But it requires bigger refactoring.

**Comment by nvoxland** [comment]
> yes, sure, I want to test the snapshot. 

If you hadn't found the build artifacts, they are best gotten from `liquibase-artifacts-added-threadlocal-scopemanager` from https://github.com/liquibase/liquibase/actions/runs/3070558330#artifacts 

> P.S. Why not make thread-safe implementation the default one? It doesn't make anything worse.

Since people use Liquibase in all sorts of ways, and this new ThreadLocal code is pretty new I'm not ready to make it the default yet. After we get some feedback and real-world usage of this new class we may very well make it the default.

> NB! I would say using ThreadLocal is rather a hack. A really God solution would be to avoid using static methods like Scope.getCurrentScope() at all. A new I stance of Scope should be created in constructor `new LiquiBase()`. But it requires bigger refactoring.

It is a hack/workaround, but part of the problem we are trying to solve is actually to move away from that `liquibase.Liquibase` facade. As the functionality of liquibase has expanded, it's too complex to fit that into the style we had for the Liquibase class.  It's also too static/pre-compiled as we try to more easily support extensions that add functionality like new commands and new arguments to existing commands. So the Scope class in general is there to act like a "arguments to a function" in a world where we don't know what those arguments even can be. 

I thought a bit about trying to pass Scope along to every single function call, but that seemed painful from a coding standpoint vs. having it available as a singleton or thread or whatever makes sense for your app. Plus, I couldn't find a great syntax setup that way that supported the "variables should fall out of scope" cleanly either. So it's trying to be a bit of a middle ground to get the dynamic parameter support we need while still working with the java syntax available.

**Comment by asolntsev** [comment]
@nvoxland I don't want to abandon `new LiquiBase()`. We do use it in our projects. 
For example, we have integration tests that run in parallel and run LiquiBase with different contexts - just to be sure that all DB scripts are valid. 

I would rather say that `Scope` class has too many responsibilities. 
1. It holds the current changelog ID - this seems to be the right usage, and should be non-static.
2. it serves as a dependency injection (like `Scope.getCurrentScope().getLog(getClass())`and `Scope.getCurrentScope().getUI()`) - and this should be moved out of `Scope`. :) For example, a common dependency injection mechanism could be used instead - like Dagger, Guice or even a simple home-made DI.
3. Most of usages is `Scope.getCurrentScope().getLog(getClass())` - and this could be replaced by a usual (de-facto standard) SLF4J logging.

**Comment by loicoudot** [comment]
@asolntsev With the help of #3240 i've added this line of code `Scope.setScopeManager(new ThreadLocalScopeManager());` in my project and Liquibase 4.20 is Threadsafe again !

Thank you for that.

From my point of view the issues can be close.

**Comment by heathmorgan** [comment]
@srbaber Do you think the approach identified by @loicoudot  in https://github.com/liquibase/liquibase/issues/2966#issuecomment-1492568131 would work for us to avoid having to use the external locks you implemented, at least until it "just works" like @asolntsev comments (and I whole-heartedly agree with!)

**Comment by relandrew** [comment]
We ran into this issue a bunch. Since we don’t actually need our liquibase migrations to run in parallel on separate databases, we wrapped the calls to liquibase with a lock, and that helped a lot. (Though we do sometimes still see some errors about scope.) The code for our workaround is something basic like `private static Object liquibaseLock = new Object(); … synchronized (liquibaseLock) { doLiquibaseStuff() }`.

**Comment by filipelautert** [comment]
This was addressed in two stages:
1. PR #3240 added `ThreadLocalScopeManager` as an opt-in workaround
2. PR #6446 made it work out of the box — `Scope` now uses `InheritableThreadLocal<ScopeManager>`, so each thread automatically gets its own scope manager. No manual `setScopeManager()` call needed.

Liquibase is now thread-safe for multi-threaded usage without any workarounds. Closing this out — if you're still hitting scope issues on a recent version, feel free to re-open.
